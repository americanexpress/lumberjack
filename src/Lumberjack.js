/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

const { Console } = require('console');
const { format } = require('util');

function loggerFunctionBuilder({
  formatter, console, beforeWrite, afterWrite,
}) {
  // so create the method either with or without the calls to before/afterWrite, rather than
  // checking if the methods were provided every console write
  let writeToConsole = (consoleMethod) => (line) => consoleMethod(line);

  if (beforeWrite && afterWrite) {
    writeToConsole = (consoleMethod) => (line) => {
      beforeWrite();
      consoleMethod(line);
      afterWrite();
    };
  } else if (beforeWrite) {
    writeToConsole = (consoleMethod) => (line) => {
      beforeWrite();
      consoleMethod(line);
    };
  } else if (afterWrite) {
    writeToConsole = (consoleMethod) => (line) => {
      consoleMethod(line);
      afterWrite();
    };
  }

  return (level) => {
    const writeToConsoleMethod = writeToConsole(console[level]);

    // The performance of .apply and the spread operator seems on par in V8
    // 6.3 but the spread operator, unlike .apply(), pushes the elements
    // onto the stack. That is, it makes stack overflows more likely.
    // https://github.com/nodejs/node/blob/master/lib/console.js#L137-L139
    // TODO: use apply instead of spread to decrease chances of stack overflows?
    return (...args) => {
      const line = formatter(level, ...args);

      if (line === null) {
        return;
      }

      writeToConsoleMethod(line);
    };
  };
}

export default function Lumberjack(opts = {}) {
  if (!(this instanceof Lumberjack)) {
    return new Lumberjack(opts);
  }

  const console = new Console(opts.stdout || process.stdout, opts.stderr);
  const formatter = opts.formatter || ((level, ...args) => format(...args));
  const beforeWrite = opts.beforeWrite || undefined;
  const afterWrite = opts.afterWrite || undefined;

  if (typeof formatter !== 'function') {
    throw new TypeError(`formatter must be a function (given "${typeof formatter}")`);
  }

  if (typeof beforeWrite !== 'function' && typeof beforeWrite !== 'undefined') {
    throw new TypeError(`beforeWrite must be a function (given "${typeof beforeWrite}")`);
  }

  if (typeof afterWrite !== 'function' && typeof afterWrite !== 'undefined') {
    throw new TypeError(`afterWrite must be a function (given "${typeof afterWrite}")`);
  }

  const getLoggerFunction = loggerFunctionBuilder({
    formatter, console, beforeWrite, afterWrite,
  });
  this.error = getLoggerFunction('error');
  this.warn = getLoggerFunction('warn');
  this.info = getLoggerFunction('info');
  this.log = getLoggerFunction('log');

  return this;
}
