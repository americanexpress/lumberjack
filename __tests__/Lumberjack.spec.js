/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import util from 'util';
import { Writable } from 'stream';

import Lumberjack from '../src/Lumberjack';

jest.mock('util', () => ({ format: jest.fn() }));

describe('Lumberjack', () => {
  function createStream({ write } = {}) {
    const stream = new Writable({
      // required by Console
      // https://github.com/nodejs/node/blob/v12.0.0/lib/internal/console/constructor.js#L83
      write: write || jest.fn(),
    });
    jest.spyOn(stream, 'write');
    // https://github.com/nodejs/node/blob/v12.0.0/lib/internal/console/constructor.js#L236
    jest.spyOn(stream, 'once');
    // https://github.com/nodejs/node/blob/v12.0.0/lib/internal/console/constructor.js#L246
    jest.spyOn(stream, 'removeListener');
    return stream;
  }

  describe('constructor', () => {
    it('is a constructor', () => {
      const logger = new Lumberjack();
      expect(logger instanceof Lumberjack).toBe(true);
    });

    it('enforces usage of `new`', () => {
      const logger = Lumberjack();
      expect(logger instanceof Lumberjack).toBe(true);
    });

    describe('stdout option', () => {
      it('throws if not a stream', () => {
        expect(() => new Lumberjack({ stdout: true })).toThrowErrorMatchingSnapshot();
      });
    });

    describe('stderr option', () => {
      it('throws if not a stream', () => {
        expect(() => new Lumberjack({ stderr: true })).toThrowErrorMatchingSnapshot();
      });
    });

    describe('formatter option', () => {
      it('defaults to util.format', () => {
        const logger = new Lumberjack();
        util.format.mockClear();
        logger.log('something');
        // jest uses something that calls util.format (probably console.log)
        // so we can't test for the number of times it's been called (desire 1)
        expect(util.format).toHaveBeenCalled();
        expect(util.format).toHaveBeenCalledWith('something');
      });

      it('throws an error if not a function', () => {
        expect(() => new Lumberjack({ formatter: true })).toThrowErrorMatchingSnapshot();
      });
    });

    describe('beforeWrite option', () => {
      it('throws an error if provided but is not a function', () => {
        expect(() => new Lumberjack({ beforeWrite: true })).toThrowErrorMatchingSnapshot();
      });
    });

    describe('afterWrite option', () => {
      it('throws an error if provided but is not a function', () => {
        expect(() => new Lumberjack({ afterWrite: true })).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe('error', () => {
    const stdout = createStream();
    const stderr = createStream();
    const formatter = jest.fn(() => 'error string');
    const logger = new Lumberjack({ stdout, stderr, formatter });

    beforeEach(() => {
      stdout.write.mockClear();
      stderr.write.mockClear();
      formatter.mockClear();
    });

    it('is formatted by the formatter option', () => {
      const a = 1;
      const b = 'two';
      const c = function three() {};
      logger.error(a, b, c);
      expect(formatter).toHaveBeenCalledTimes(1);
      expect(formatter.mock.calls[0]).toEqual(['error', a, b, c]);
    });

    it('is written to the stderr stream option', () => {
      logger.error();
      expect(stderr.write).toHaveBeenCalledTimes(1);
      expect(stderr.write.mock.calls[0][0]).toEqual('error string\n');
    });

    it('is not written to the stdout stream option', () => {
      logger.error();
      expect(stdout.write).not.toHaveBeenCalled();
    });

    it('skips writing to the stderr stream if formatter returned null', () => {
      const skipperLogger = new Lumberjack({ stdout, stderr, formatter: () => null });
      skipperLogger.error('skipping');
      expect(stderr.write).not.toHaveBeenCalled();
    });

    it('calls beforeWrite before writing to stderr', () => {
      const order = [];
      const beforeWritePush = jest.fn(() => { order.push('beforeWrite'); });
      const stderrPush = createStream({ write: jest.fn(() => { order.push('stderr'); }) });
      const beforeLogger = new Lumberjack({
        stdout,
        stderr: stderrPush,
        formatter: (v) => v,
        beforeWrite: beforeWritePush,
      });
      beforeLogger.error('text');
      expect(beforeWritePush).toHaveBeenCalledTimes(1);
      expect(stderrPush.write).toHaveBeenCalled();
      expect(order).toEqual(['beforeWrite', 'stderr']);
    });

    it('calls afterWrite after writing to stderr', () => {
      const order = [];
      const afterWritePush = jest.fn(() => { order.push('afterWrite'); });
      const stderrPush = createStream({ write: jest.fn(() => { order.push('stderr'); }) });
      const afterLogger = new Lumberjack({
        stdout,
        stderr: stderrPush,
        formatter: (v) => v,
        afterWrite: afterWritePush,
      });
      afterLogger.error('text');
      expect(afterWritePush).toHaveBeenCalledTimes(1);
      expect(stderrPush.write).toHaveBeenCalled();
      expect(order).toEqual(['stderr', 'afterWrite']);
    });

    it('calls both beforeWrite and afterWrite', () => {
      const order = [];
      const beforeWritePush = jest.fn(() => { order.push('beforeWrite'); });
      const afterWritePush = jest.fn(() => { order.push('afterWrite'); });
      const stderrPush = createStream({ write: jest.fn(() => { order.push('stderr'); }) });
      const beforeAndAfterLogger = new Lumberjack({
        stdout,
        stderr: stderrPush,
        formatter: (v) => v,
        beforeWrite: beforeWritePush,
        afterWrite: afterWritePush,
      });
      beforeAndAfterLogger.error('text');
      expect(afterWritePush).toHaveBeenCalledTimes(1);
      expect(stderrPush.write).toHaveBeenCalled();
      expect(order).toEqual(['beforeWrite', 'stderr', 'afterWrite']);
    });
  });

  describe('warn', () => {
    const stdout = createStream();
    const stderr = createStream();
    const formatter = jest.fn(() => 'warn string');
    const logger = new Lumberjack({ stdout, stderr, formatter });

    beforeEach(() => {
      stdout.write.mockClear();
      stderr.write.mockClear();
      formatter.mockClear();
    });

    it('is formatted by the formatter option', () => {
      const a = 1;
      const b = 'two';
      const c = function three() {};
      logger.warn(a, b, c);
      expect(formatter).toHaveBeenCalledTimes(1);
      expect(formatter.mock.calls[0]).toEqual(['warn', a, b, c]);
    });

    it('is written to the stderr stream option', () => {
      logger.warn();
      expect(stderr.write).toHaveBeenCalledTimes(1);
      expect(stderr.write.mock.calls[0][0]).toEqual('warn string\n');
    });

    it('is not written to the stdout stream option', () => {
      logger.warn();
      expect(stdout.write).not.toHaveBeenCalled();
    });

    it('skips writing to the stderr stream if formatter returned null', () => {
      const skipperLogger = new Lumberjack({ stdout, stderr, formatter: () => null });
      skipperLogger.warn('skipping');
      expect(stderr.write).not.toHaveBeenCalled();
    });

    it('calls beforeWrite before writing to stderr', () => {
      const order = [];
      const beforeWritePush = jest.fn(() => { order.push('beforeWrite'); });
      const stderrPush = createStream({ write: jest.fn(() => { order.push('stderr'); }) });
      const beforeLogger = new Lumberjack({
        stdout,
        stderr: stderrPush,
        formatter: (v) => v,
        beforeWrite: beforeWritePush,
      });
      beforeLogger.warn('text');
      expect(beforeWritePush).toHaveBeenCalledTimes(1);
      expect(stderrPush.write).toHaveBeenCalled();
      expect(order).toEqual(['beforeWrite', 'stderr']);
    });

    it('calls afterWrite after writing to stderr', () => {
      const order = [];
      const afterWritePush = jest.fn(() => { order.push('afterWrite'); });
      const stderrPush = createStream({ write: jest.fn(() => { order.push('stderr'); }) });
      const afterLogger = new Lumberjack({
        stdout,
        stderr: stderrPush,
        formatter: (v) => v,
        afterWrite: afterWritePush,
      });
      afterLogger.warn('text');
      expect(afterWritePush).toHaveBeenCalledTimes(1);
      expect(stderrPush.write).toHaveBeenCalled();
      expect(order).toEqual(['stderr', 'afterWrite']);
    });

    it('calls both beforeWrite and afterWrite', () => {
      const order = [];
      const beforeWritePush = jest.fn(() => { order.push('beforeWrite'); });
      const afterWritePush = jest.fn(() => { order.push('afterWrite'); });
      const stderrPush = createStream({ write: jest.fn(() => { order.push('stderr'); }) });
      const beforeAndAfterLogger = new Lumberjack({
        stdout,
        stderr: stderrPush,
        formatter: (v) => v,
        beforeWrite: beforeWritePush,
        afterWrite: afterWritePush,
      });
      beforeAndAfterLogger.warn('text');
      expect(afterWritePush).toHaveBeenCalledTimes(1);
      expect(stderrPush.write).toHaveBeenCalled();
      expect(order).toEqual(['beforeWrite', 'stderr', 'afterWrite']);
    });
  });

  describe('info', () => {
    const stdout = createStream();
    const formatter = jest.fn(() => 'info string');
    const logger = new Lumberjack({ stdout, formatter });

    beforeEach(() => {
      stdout.write.mockClear();
      formatter.mockClear();
    });

    it('is formatted by the formatter option', () => {
      const a = 1;
      const b = 'two';
      const c = function three() {};
      logger.info(a, b, c);
      expect(formatter).toHaveBeenCalledTimes(1);
      expect(formatter.mock.calls[0]).toEqual(['info', a, b, c]);
    });

    it('is written to the stdout stream option', () => {
      logger.info();
      expect(stdout.write).toHaveBeenCalledTimes(1);
      expect(stdout.write.mock.calls[0][0]).toEqual('info string\n');
    });

    it('skips writing to the stdout stream if formatter returned null', () => {
      const skipperLogger = new Lumberjack({ stdout, formatter: () => null });
      skipperLogger.error('skipping');
      expect(stdout.write).not.toHaveBeenCalled();
    });

    it('calls beforeWrite before writing to stderr', () => {
      const order = [];
      const beforeWritePush = jest.fn(() => { order.push('beforeWrite'); });
      const stdoutPush = createStream({ write: jest.fn(() => { order.push('stdout'); }) });
      const beforeLogger = new Lumberjack({
        stdout: stdoutPush,
        formatter: (v) => v,
        beforeWrite: beforeWritePush,
      });
      beforeLogger.info('text');
      expect(beforeWritePush).toHaveBeenCalledTimes(1);
      expect(stdoutPush.write).toHaveBeenCalled();
      expect(order).toEqual(['beforeWrite', 'stdout']);
    });

    it('calls afterWrite after writing to stdout', () => {
      const order = [];
      const afterWritePush = jest.fn(() => { order.push('afterWrite'); });
      const stdoutPush = createStream({ write: jest.fn(() => { order.push('stdout'); }) });
      const afterLogger = new Lumberjack({
        stdout: stdoutPush,
        formatter: (v) => v,
        afterWrite: afterWritePush,
      });
      afterLogger.info('text');
      expect(afterWritePush).toHaveBeenCalledTimes(1);
      expect(stdoutPush.write).toHaveBeenCalled();
      expect(order).toEqual(['stdout', 'afterWrite']);
    });

    it('calls both beforeWrite and afterWrite', () => {
      const order = [];
      const beforeWritePush = jest.fn(() => { order.push('beforeWrite'); });
      const afterWritePush = jest.fn(() => { order.push('afterWrite'); });
      const stdoutPush = createStream({ write: jest.fn(() => { order.push('stdout'); }) });
      const beforeAndAfterLogger = new Lumberjack({
        stdout: stdoutPush,
        formatter: (v) => v,
        beforeWrite: beforeWritePush,
        afterWrite: afterWritePush,
      });
      beforeAndAfterLogger.info('text');
      expect(afterWritePush).toHaveBeenCalledTimes(1);
      expect(stdoutPush.write).toHaveBeenCalled();
      expect(order).toEqual(['beforeWrite', 'stdout', 'afterWrite']);
    });
  });

  describe('log', () => {
    const stdout = createStream();
    const formatter = jest.fn(() => 'log string');
    const logger = new Lumberjack({ stdout, formatter });

    beforeEach(() => {
      stdout.write.mockClear();
      formatter.mockClear();
    });

    it('is formatted by the formatter option', () => {
      const a = 1;
      const b = 'two';
      const c = function three() {};
      logger.log(a, b, c);
      expect(formatter).toHaveBeenCalledTimes(1);
      expect(formatter.mock.calls[0]).toEqual(['log', a, b, c]);
    });

    it('is written to the stdout stream option', () => {
      logger.log();
      expect(stdout.write).toHaveBeenCalledTimes(1);
      expect(stdout.write.mock.calls[0][0]).toEqual('log string\n');
    });

    it('skips writing to the stdout stream if formatter returned null', () => {
      const skipperLogger = new Lumberjack({ stdout, formatter: () => null });
      skipperLogger.error('skipping');
      expect(stdout.write).not.toHaveBeenCalled();
    });

    it('calls beforeWrite before writing to stderr', () => {
      const order = [];
      const beforeWritePush = jest.fn(() => { order.push('beforeWrite'); });
      const stdoutPush = createStream({ write: jest.fn(() => { order.push('stdout'); }) });
      const beforeLogger = new Lumberjack({
        stdout: stdoutPush,
        formatter: (v) => v,
        beforeWrite: beforeWritePush,
      });
      beforeLogger.log('text');
      expect(beforeWritePush).toHaveBeenCalledTimes(1);
      expect(stdoutPush.write).toHaveBeenCalled();
      expect(order).toEqual(['beforeWrite', 'stdout']);
    });

    it('calls afterWrite after writing to stdout', () => {
      const order = [];
      const afterWritePush = jest.fn(() => { order.push('afterWrite'); });
      const stdoutPush = createStream({ write: jest.fn(() => { order.push('stdout'); }) });
      const afterLogger = new Lumberjack({
        stdout: stdoutPush,
        formatter: (v) => v,
        afterWrite: afterWritePush,
      });
      afterLogger.log('text');
      expect(afterWritePush).toHaveBeenCalledTimes(1);
      expect(stdoutPush.write).toHaveBeenCalled();
      expect(order).toEqual(['stdout', 'afterWrite']);
    });

    it('calls both beforeWrite and afterWrite', () => {
      const order = [];
      const beforeWritePush = jest.fn(() => { order.push('beforeWrite'); });
      const afterWritePush = jest.fn(() => { order.push('afterWrite'); });
      const stdoutPush = createStream({ write: jest.fn(() => { order.push('stdout'); }) });
      const beforeAndAfterLogger = new Lumberjack({
        stdout: stdoutPush,
        formatter: (v) => v,
        beforeWrite: beforeWritePush,
        afterWrite: afterWritePush,
      });
      beforeAndAfterLogger.log('text');
      expect(afterWritePush).toHaveBeenCalledTimes(1);
      expect(stdoutPush.write).toHaveBeenCalled();
      expect(order).toEqual(['beforeWrite', 'stdout', 'afterWrite']);
    });
  });
});
