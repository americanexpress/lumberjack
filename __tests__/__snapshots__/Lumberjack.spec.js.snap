// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lumberjack constructor afterWrite option throws an error if provided but is not a function 1`] = `"afterWrite must be a function (given "boolean")"`;

exports[`Lumberjack constructor beforeWrite option throws an error if provided but is not a function 1`] = `"beforeWrite must be a function (given "boolean")"`;

exports[`Lumberjack constructor formatter option throws an error if not a function 1`] = `"formatter must be a function (given "boolean")"`;

exports[`Lumberjack constructor stderr option throws if not a stream 1`] = `"Console expects a writable stream instance for stderr"`;

exports[`Lumberjack constructor stdout option throws if not a stream 1`] = `"Console expects a writable stream instance for stdout"`;
